### PR TITLE
Move cb_sample to be class based

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: vowpalwabbit/cb_sample.cc [hunk 1]
```diff
@@ -93,8 +102,7 @@ base_learner* cb_sample_setup(options_i& options, vw& all)
   if (!cb_sample_option)
     return nullptr;
 
-  data->seed_state = all.random_seed;
-
+  auto data = scoped_calloc_or_throw<cb_sample_data>(all.random_seed);
```
Comment (rajan-chari) [inline vowpalwabbit/cb_sample.cc]: Peter's PR https://github.com/VowpalWabbit/vowpal_wabbit/pull/2069 creates a shared random state object.  Perhaps we should be using that here?

File: vowpalwabbit/cb_sample.cc [hunk 2]
```diff
@@ -8,82 +8,91 @@ using namespace LEARNER;
 using namespace VW;
 using namespace VW::config;
 
+namespace VW
+{
 struct cb_sample_data
```
Comment (rajan-chari) [inline vowpalwabbit/cb_sample.cc]: Some explanation of what this reduction does will be useful.

File: vowpalwabbit/cb_sample.cc [hunk 3]
```diff
@@ -8,82 +8,91 @@ using namespace LEARNER;
...
+void learn_or_predict(cb_sample_data &data, multi_learner &base, multi_ex &examples)
+{
+  data.learn_or_predict<is_learn>(base, examples);
 }
 
-base_learner* cb_sample_setup(options_i& options, vw& all)
+base_learner *cb_sample_setup(options_i &options, vw &all)
 {
-  auto data = scoped_calloc_or_throw<cb_sample_data>();
   bool cb_sample_option = false;
```
Comment (rajan-chari) [inline vowpalwabbit/cb_sample.cc]: I don't see where this is set to true?  Does this setup function always return nullptr?

File: vowpalwabbit/cb_sample.h [hunk 4]
```diff
@@ -1,3 +1,5 @@
 #pragma once
-
+
+#include "reductions_fwd.h"
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.h]: Yes, this header brings forward definitions used in the below declaration

File: vowpalwabbit/cb_sample.cc (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.cc]: Yes, I'd love to consume that object here!

File: vowpalwabbit/cb_sample.cc (hunk 3 above)
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.cc]: This is set to true by the option parser on line 90

File: vowpalwabbit/cb_sample.cc (hunk 2 above)
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.cc]: Sounds good, I'll add a doc comment

File: vowpalwabbit/cb_sample.cc [hunk 5]
```diff
@@ -8,82 +8,91 @@ using namespace LEARNER;
...
-  uint64_t seed_state;
-};
+  explicit cb_sample_data(uint64_t seedState) : _seed_state(seedState) {}
 
-template <bool is_learn>
-void learn_or_predict(cb_sample_data& data, multi_learner& base, multi_ex& examples)
-{
-  multiline_learn_or_predict<is_learn>(base, examples, examples[0]->ft_offset);
+  template <bool is_learn>
+  void learn_or_predict(multi_learner &base, multi_ex &examples)
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.cc]: We don't make any other learn_or_predict functions inline, why this one?

[the remark this answers is not in the document]
Comment (rajan-chari) [reply vowpalwabbit/cb_sample.cc]: Good question.  By adding classes we now have one more method.  Perhaps there is more opportunity to inline now.

[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/cb_sample.cc]: I agree, the functions that are required to get learn/predict happening should be easily inlined

**Comment by rajan-chari** [comment]
Merged thanks.

**Pull request merged**